README.md
# ios_app
The Lane Breach iOS app allows users to report bicycle lane blockages to San Francisco's 311 service. 

The existing 311 mobile app already offers this functionality, but it's clunky and not intuitive. 
This app offers a streamlined experience and also provides users with a visualization of bike infrastructure
and lane blockages reported by all users city-wide, as well as a summary of all incidents reported by the current user.

## Getting Started
To build this app, you'll need a Mac with a recent-ish version of OS X and Xcode (the original version of this app was 
developed with Xcode 10.1 running on OS X High Sierra). This app is written in Swift 4.

## Contributions
We welcome contributions! All of our contributors are expected to follow the rules outlined in our Code of Conduct. 
Please create a separate pull request against master for each new feature/bug-fix. We will try to use Github issues 
whenever possible to describe tasks that we need to do.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 1.2.0 (1)
- twitter screen - added progress indicator when the screen is initially loading
- settings screen - added "How to use Lane Breach" action which takes the user to a web help screen; dismiss keyboard if user touches anywhere on the screen
- new report screen - make sure that we keep the picker in sync with the current category (this is only used after previously-uploaded report's category is reset)
- fix bug where the last character of a user-entered description would not be submitted to 311
- shrink images more than 500px wide to make 'em smaller for uploading
- added imageSourceCamera (true/false) and imageTimeDelta (natural log of time since the picture was taken in seconds) to NetworkManagerSuccess events
- added user notifications after a report is uploaded (including a new optional permission)
- new app icons

## 1.1.0 (1)
- permissions screen - new screen that replaces random permission popups during app launch and makes it easier (in theory) to recover from a revoked permission while the app is running
- new report screen - added 1/2/4X zoom function; added white shadows to black icons so they're easier to see at night; hide cursor on text field when category picker is showing; moved Other to the top of the list.
- twitter screen - new screen that shows @EverySF311Bike's feed with the user's reports marked with a star
- always prepend [Other] before submitted descriptions if a category is not picked
- switch to custom Fabric events to report successful report delivery to 311

## 1.0.0 (12)
- change deployment target to iOS 11 (from 11.4)
- new report screen - don't allow changing the flash mode if the device doesn't support it otherwise start with the flash off; (grudgingly) accept the 3rd location if our accuracy requirement isn't met as long as it's 100m or better.
- settings screen - change the server switch action to be long-press; tweak text/positions to look better on smaller screens.

## 1.0.0 (10)
- new report screen - update the location icon based on having captured a valid location or the image having a location

## 1.0.0 (9)
- updated map style to point to the latest iOS mapbox style
- settings screen - fixed bug where the fields were not updated unless the user pressed return after changing each one

## 1.0.0 (8)
- disable token -> service request ID check since it never succeeds and just makes the user wait 2+ seconds more when submitting
- tweak map annotation
- use "background tasks" to improve uploading if the user backgrounds the app in the middle of the action
- treat HTTP 4xx errors as permanent fatal errors (since we get 400 if 311 detects a dupe POST even if we think that the first one failed)
- improve location collection algorithm
- better network error codes for showing to users
- report various stats to Crashlytics as non-fatal errors
- added Crashlytics to track app crashes

## 1.0.0 (7)
- one needs to talk to the prod server when using the prod key

## 1.0.0 (6)
- remember which hints are hidden and add way to restore 'em all. Also added "Commuter Shuttle" as a category.
- added prod API key and a super fancy way to switch between dev and prod
- update map style and add legend
- settings screen - added name/phone number (optional) fields to be submitted with the 311 report if supplied
- new report screen - fix bug where keyboard entry on the settings screen would screw up the vertical position of the data entry box on this screen.


blfa/NetworkManager.swift
//
//  NetworkManager.swift
//  blfa
//
//  Created by Dale Low on 2/15/19.
//  Copyright © 2019 Dale Low. All rights reserved.
//

import AWSS3
import CoreLocation
import Foundation

public class NetworkManagerError: NSError {
    enum ErrorKind: Int {
        case networkGeneral = -1
        case networkBadHTTPStatus = -2
        case networkResponseMalformed = -3
        case networkJsonDecodeFailure = -4
        case dataTaskNullDataOrError = -5
        case imageUploadFailed = -6
        case missingServiceRequestId = -7
        case missingServiceRequestIdAndToken = -8
    }
    
    let kind: ErrorKind
    
    init(_ kind: ErrorKind, description: String? = nil, httpStatusCode: Int? = nil, domain: String? = nil, function: String = #function, line: Int = #line) {
        self.kind = kind
        
        var localizedDescription: String
        if let description = description {
            localizedDescription = "\(function):\(line) - \(description)"
        } else {
            localizedDescription = "\(function):\(line)"
        }
        
        var errorDomain: String
        if let domain = domain {
            errorDomain = domain
        } else {
            if UserDefaults.standard.bool(forKey: kUserDefaultsUsingDevServerKey) {
                errorDomain = "DevNetworkManagerError"
            } else {
                errorDomain = "NetworkManagerError"
            }
        }

        // use HTTP status as NSError code, otherwise use Kind
        var code: Int
        if let httpStatusCode = httpStatusCode {
            code = httpStatusCode
        } else {
            code = kind.rawValue
        }
        
        super.init(domain: errorDomain, code: code, userInfo: [NSLocalizedDescriptionKey: localizedDescription])
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}

public class MockNetworkManagerError: NetworkManagerError {
    override init(_ kind: ErrorKind, description: String? = nil, httpStatusCode: Int? = nil, domain: String? = nil, function: String = #function, line: Int = #line) {
        super.init(kind, description: description, httpStatusCode: httpStatusCode, domain: "MockNetworkManagerError")
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}

class NetworkManager {
    let baseDomainDevelopment = "mobile311-dev.sfgov.org"
    let baseDomainProduction = "mobile311.sfgov.org"
    var mockUploadReportCount = -1

    static let shared = NetworkManager()
    var debugLastHttpPost: String?
    var uploadReportTaskId: UIBackgroundTaskIdentifier?
    var uploadReportCompletion: ((_ serviceRequestId: String?, _ token: String?, _ error: NetworkManagerError?) -> Void)?
    
    func updateTabBarStyleForCurrentServer(vc: UIViewController) {
        vc.tabBarController?.tabBar.barTintColor = UserDefaults.standard.bool(forKey: kUserDefaultsUsingDevServerKey) ? UIColor.red : UIColor.lightGray
        vc.tabBarController?.tabBar.unselectedItemTintColor = UIColor.black
        vc.tabBarController?.tabBar.tintColor = UIColor.blue
    }

    private func uploadImage(with data: Data, filename: String, completion: @escaping (Error?) -> Void) {
        let expression = AWSS3TransferUtilityUploadExpression()
//        expression.progressBlock = progressBlock
        
        let completionHandler: AWSS3TransferUtilityUploadCompletionHandlerBlock = { (task, error) -> Void in
            DispatchQueue.main.async(execute: {
                if let error = error {
                    print("completionHandler failed with error: \(error)")
                    completion(error)
                }
//                else if (self.progressView.progress != 1.0) {
//                    print("Error: Failed - Likely due to invalid region / filename")
//                }
                else {
                    print("completionHandler success")
                    completion(nil)
                }
            })
        }
        
        let transferUtility = AWSS3TransferUtility.s3TransferUtility(forKey: "USWest1S3TransferUtility")
        transferUtility.uploadData(
            data,
            bucket: "lane-breach",
            key: "311-sf/temp-images/\(filename).png",
            contentType: "image/png",
            expression: expression,
            completionHandler: completionHandler).continueWith { (task) -> AnyObject? in
                if let error = task.error {
                    print("uploadData failed with error: \(error)")
                    completion(error)
                }
                
                if let _ = task.result {
                    print("uploadData starting")
                    // Do something with uploadTask.
                }
                
                return nil;
        }
    }

    func getServiceRequestId(from token: String, completion: @escaping (String?, Error?) -> Void) {
        // now do a GET to translate the token into a service_request_id
        // ex: http://mobile311-dev.sfgov.org/open311/v2/tokens/5bc6c0f5ff031d6f5b335df0.json
        let baseDomain = UserDefaults.standard.bool(forKey: kUserDefaultsUsingDevServerKey) ? baseDomainDevelopment : baseDomainProduction;
        let url = URL(string: "http://\(baseDomain)/open311/v2/tokens/\(token).json")!
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        let task = URLSession.shared.dataTask(with: request) { data, response, error in
            DispatchQueue.main.async {
                guard let data = data, error == nil else {                                                 // check for fundamental networking error
                    completion(nil, NetworkManagerError(.networkGeneral, description: error != nil ? "error='\(error!)'" : "error='no data'"))
                    return
                }
                
                // check for 200
                if let httpStatus = response as? HTTPURLResponse, httpStatus.statusCode != 200 {           // check for http errors
                    print("statusCode should be 200, but is \(httpStatus.statusCode)")
                    print("response = \(httpStatus)")
                    completion(nil, NetworkManagerError(.networkBadHTTPStatus, httpStatusCode: httpStatus.statusCode))
                    return
                }
                
                if let responseString = String(data: data, encoding: .utf8) {
                    // looks like: responseString = [{"token":"5bc6c0f5ff031d6f5b335df0"}]
                    print("responseString2 = \(responseString)")
                    
                    let json = try? JSONSerialization.jsonObject(with: data, options: [])
                    if let dictionary = json as? [[String: Any]] {
                        if let serviceRequestId = dictionary[0]["service_request_id"] as? String {
                            print("serviceRequestId: \(serviceRequestId)")
                            completion(serviceRequestId, nil)
                        } else {
                            completion(nil, NetworkManagerError(.missingServiceRequestId))
                        }
                    } else {
                        completion(nil, NetworkManagerError(.networkJsonDecodeFailure))
                    }
                }
            }
        }
        task.resume()
    }

    func mockUploadReport(progressMessage: @escaping (String) -> Void,
                          completion: @escaping (_ serviceRequestId: String?, _ token: String?, _ error: NetworkManagerError?) -> Void) {
        
        if mockUploadReportCount == 7 {
            mockUploadReportCount = 0
        } else {
            mockUploadReportCount += 1
        }
        
        progressMessage("Uploading image")
        DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
            if self.mockUploadReportCount == 0 {
                completion(nil, nil, MockNetworkManagerError(.imageUploadFailed, description: "error='mock image upload failed'"))
            } else {
                progressMessage("Uploading details")
                DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
                    switch self.mockUploadReportCount {
                    case 1:
                        completion(nil, nil, MockNetworkManagerError(.dataTaskNullDataOrError, description: "error='mock error'"))
                    case 2:
                        completion(nil, nil, MockNetworkManagerError(.networkBadHTTPStatus, httpStatusCode: 400))
                    case 3:
                        completion(nil, nil, MockNetworkManagerError(.missingServiceRequestIdAndToken))
                    case 4:
                        completion(nil, nil, MockNetworkManagerError(.networkJsonDecodeFailure))
                    case 5:
                        completion(nil, nil, MockNetworkManagerError(.networkResponseMalformed))
                    case 6:
                        completion("123", nil, nil)
                    case 7:
                        completion(nil, "456", nil)
                    default:
                        assert(false)
                    }
                }
            }
        }
    }
    
    func beginUploadReportTask() {
        let taskId = UIApplication.shared.beginBackgroundTask(expirationHandler: {
            self.endUploadReportTask()
        })
        
        if taskId != UIBackgroundTaskInvalid {
            print("*** starting background task with ID \(taskId)")
            self.uploadReportTaskId = taskId
        } else {
            print("*** failed to start background task")
            self.uploadReportTaskId = nil
        }
    }
    
    func endUploadReportTask() {
        if let uploadReportTaskId = uploadReportTaskId {
            print("*** ending background task with ID \(uploadReportTaskId)")
            UIApplication.shared.endBackgroundTask(uploadReportTaskId)
            self.uploadReportTaskId = nil
        }
    }
    
    func reportCompletionAndEndUploadReportTask(_ serviceRequestId: String?, _ token: String?, _ error: NetworkManagerError?) {
        if let uploadReportCompletion = uploadReportCompletion {
            uploadReportCompletion(serviceRequestId, token, error)
            self.uploadReportCompletion = nil
        } else {
            assertionFailure()
        }
        
        self.endUploadReportTask()
    }
    
    // credit: https://stackoverflow.com/questions/31314412/how-to-resize-image-in-swift
    func scaledImage(with image: UIImage, scaledTo newSize: CGSize) -> UIImage? {
        UIGraphicsBeginImageContextWithOptions(newSize, false, 1.0)
        image.draw(in: CGRect(x: 0, y: 0, width: newSize.width, height: newSize.height))
        let newImage = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        return newImage
    }
    
    func uploadReport(image: UIImage, filename: String, location: CLLocationCoordinate2D,
                      emailAddress: String?, fullName: String?, phoneNumber: String?,
                      category: String, description: String,
                      progressMessage: @escaping (String) -> Void,
                      completion: @escaping (_ serviceRequestId: String?, _ token: String?, _ error: NetworkManagerError?) -> Void) {
        
        if AppDelegate.getMockTestEnable(for: .useMockUpload) {
            mockUploadReport(progressMessage: progressMessage, completion: completion)
            return
        }
        
        print("original size: \(UIImagePNGRepresentation(image)!.count) bytes")
        
        // keep image size down
        var imageToUpload = image
        if image.size.width > 500 /* arbitrary */ {
            let scaleFactor = 500.0/image.size.width
            if let image = scaledImage(with: image, scaledTo: CGSize(width: image.size.width*scaleFactor, height: image.size.height*scaleFactor)) {
                imageToUpload = image

                print("scaled size: \(UIImagePNGRepresentation(image)!.count) bytes")
            }
        }
        
        uploadReportCompletion = completion
        beginUploadReportTask()
        progressMessage("Uploading image")
        self.uploadImage(with: UIImagePNGRepresentation(imageToUpload)!, filename: filename) { (error) in
            print("\(Date().timeIntervalSince1970) start metadata upload")
            if let error = error {
                self.reportCompletionAndEndUploadReportTask(nil, nil, NetworkManagerError(.imageUploadFailed, description: "error='\(error)'"))
                return
            }
            
            // TODO - send the mediaUrl from uploadImage()
            let mediaUrl = "https://s3-us-west-1.amazonaws.com/lane-breach/311-sf/temp-images/\(filename).png"
            
            // concatenate category with optional description when POSTing (ex: [category] <description>)
            var fullDescription: String = "[\(category)] "
            fullDescription.append(contentsOf: (description.count) != 0 ? description : "Blocked bicycle lane")
            
            var parameters = [
                "api_key": UserDefaults.standard.bool(forKey: kUserDefaultsUsingDevServerKey) ? Keys.apiKey : Keys.apiKeyProduction,
                "service_code": "5a6b5ac2d0521c1134854b01",
                "lat": String(location.latitude),
                "long": String(location.longitude),
                "media_url": mediaUrl,
                "description": fullDescription,
                "attribute[Nature_of_request]": "Blocking_Bicycle_Lane"
            ]
            
            if let emailAddress = emailAddress, emailAddress.trimmingCharacters(in: .whitespaces).count > 0 {
                parameters["email"] = emailAddress.trimmingCharacters(in: .whitespaces)
            } else {
                parameters["email"] = "bikelanessf@gmail.com"
            }
            
            if let fullName = fullName, fullName.trimmingCharacters(in: .whitespaces).count > 0 {
                var components = fullName.components(separatedBy: " ")
                if components.count > 0 {
                    parameters["first_name"] = components.removeFirst()
                    parameters["last_name"] = components.joined(separator: " ").trimmingCharacters(in: .whitespaces)
                } else {
                    parameters["first_name"] = fullName
                    parameters["last_name"] = ""
                }
            }

            if let phoneNumber = phoneNumber, phoneNumber.trimmingCharacters(in: .whitespaces).count > 0 {
                parameters["phone"] = phoneNumber.trimmingCharacters(in: .whitespaces)
            }
            
            // create the form URL-encoded string for the params
            var postString: String = ""
            var first = true
            for (key, value) in parameters {
                let escapedString = value.addingPercentEncoding(withAllowedCharacters: .urlHostAllowed)!
                
                if first {
                    first = false
                } else {
                    postString += "&"
                }
                postString += "\(key)=\(escapedString)"
            }
            self.debugLastHttpPost = postString
            
            // POST it
            let baseDomain = UserDefaults.standard.bool(forKey: kUserDefaultsUsingDevServerKey) ? self.baseDomainDevelopment : self.baseDomainProduction;
            let url = URL(string: "http://\(baseDomain)/open311/v2/requests.json")!
            var request = URLRequest(url: url)
            request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
            request.httpMethod = "POST"
            request.httpBody = postString.data(using: .utf8)
            
            progressMessage("Uploading details")
            let task = URLSession.shared.dataTask(with: request) { data, response, error in
                DispatchQueue.main.async {
                    guard let data = data, error == nil else {                                                 // check for fundamental networking error
                        self.reportCompletionAndEndUploadReportTask(nil, nil, NetworkManagerError(.dataTaskNullDataOrError, description: "error='\(String(describing: error))'"))
                        return
                    }
                    
                    // check for 201 CREATED
                    if let httpStatus = response as? HTTPURLResponse, httpStatus.statusCode != 201 {           // check for http errors
                        print("statusCode should be 201, but is \(httpStatus.statusCode)")
                        print("response = \(httpStatus)")
                        self.reportCompletionAndEndUploadReportTask(nil, nil, NetworkManagerError(.networkBadHTTPStatus, httpStatusCode: httpStatus.statusCode))
                        return
                    }
                    
                    if let responseString = String(data: data, encoding: .utf8) {
                        // looks like: responseString = [{"token":"5bc6c0f5ff031d6f5b335df0"}]
                        print("responseString = \(responseString)")
                        
                        // check if it's a token
                        let json = try? JSONSerialization.jsonObject(with: data, options: [])
                        if let dictionary = json as? [[String: Any]] {
                            if let serviceRequestId = dictionary[0]["service_request_id"] as? String {
                                print("serviceRequestId: \(serviceRequestId)")
                                self.reportCompletionAndEndUploadReportTask(serviceRequestId, nil, nil)
                            } else if let token = dictionary[0]["token"] as? String {
                                print("token: \(token)")
                                self.reportCompletionAndEndUploadReportTask(nil, token, nil)
                                
                                // need a delay to allow 311 to get a service request ID
                                // this never seems to succeed on dev or prod, so might as well complete faster and skip it
//                                DispatchQueue.main.asyncAfter(deadline: .now() + 2) {
//                                    self.getServiceRequestId(from: token, completion: { (serviceRequestId, error) in
//                                        // don't pass the error, if any, along cuz we've always got a token
//                                        self.reportCompletionAndEndUploadReportTask(serviceRequestId, token, nil)
//                                    })
//                                }
//                                return
                            } else {
                                self.reportCompletionAndEndUploadReportTask(nil, nil, NetworkManagerError(.missingServiceRequestIdAndToken))
                            }
                        } else {
                            self.reportCompletionAndEndUploadReportTask(nil, nil, NetworkManagerError(.networkJsonDecodeFailure))
                        }
                    } else {
                        self.reportCompletionAndEndUploadReportTask(nil, nil, NetworkManagerError(.networkResponseMalformed))
                    }
                    
                    print("\(Date().timeIntervalSince1970) done metadata upload")
                }
            }
            task.resume()
        }
    }
}
